kroxylicious/src/main/java/io/kroxylicious/proxy/internal/ResponseOrderer.java

```diff
@@ -0,0 +1,65 @@
+/*
+ * Copyright Kroxylicious Authors.
+ *
+ * Licensed under the Apache Software License version 2.0, available at http://www.apache.org/licenses/LICENSE-2.0
+ */
+
+package io.kroxylicious.proxy.internal;
+
+import java.util.ArrayDeque;
+import java.util.Deque;
+import java.util.HashMap;
+import java.util.Map;
+
+import io.netty.channel.ChannelDuplexHandler;
+import io.netty.channel.ChannelHandlerContext;
+import io.netty.channel.ChannelPromise;
+
+import io.kroxylicious.proxy.frame.Frame;
+
+public class ResponseOrderer extends ChannelDuplexHandler {
+
```

Comment:
Could we add unit tests exercising the case where > 1 responses are outstanding?

Reply:
Added unit tests covering basic inbound/outbound forward and these scenarios:

1. request-1 in, response-1 out
2. request-1 in, request-2 in, response-1 out, response-2 out
3. request-1 in, request-2 in, response-2 out, response-1 out (checks that the responses are re-ordered)
3. request-1 in, request-2 in, request-3 in, request-4 in response-2 out, response-4 out, response-1 out, response-3 out (checks the case where we can drain some but not all of the queued responses)

Reply:
edit: and added

request-1 in, request-2 in, request-3 in, response-2 out, response-3 out, response-1 out (checks the case where we can drain multiple pending responses in reaction to the response-1 write)

Reply:
Reading the Kafka Protocol guide it says:

> The server guarantees that on a single TCP connection, requests will be processed in the order they are sent and responses will return in that order as well. The broker's request processing allows only a single in-flight request per connection in order to guarantee this ordering.

Which would suggest that we only need to worry about a single outstanding request.

However it goes on to say 
> Note that clients can (and ideally should) use non-blocking IO to implement request pipelining and achieve higher throughput. i.e., clients can send requests even while awaiting responses for preceding requests since the outstanding requests will be buffered in the underlying OS socket buffer

Which implies multiple requests can be considered in-flight...

Have you been able to demonstrate multiple requests in flight? I only ask as supporting a single outstanding request is a lot cheaper perf wise than supporting `N`.